test/unit/puppetlabs/puppetserver/lockable_pool_test.clj

```diff
@@ -60,6 +85,27 @@
         ;; make sure we got here
         (is (true? true))))))
 
+(deftest pool-lock-is-blocking-test-until-borrows-unregistered-test
+  (let [pool (create-populated-pool 3)
+        instances (borrow-n-instances pool 2)
+        future-started? (promise)]
+    (is (= 2 (count instances)))
+    (let [lock-thread (future (deliver future-started? true)
+                              (.lock pool))]
+      @future-started?
+      (testing "pool.lock() blocks until all instances are unregistered"
+        (is (not (realized? lock-thread)))
+
+        (testing (str "other threads may successfully unregister instances "
+                      "while pool.lock() is being executed")
+          (.unregister pool (first instances))
+          (is (not (realized? lock-thread)))
+          (.unregister pool (second instances)))
+
+        @lock-thread
+        ;; make sure we got here
+        (is (true? true))))))
```

Comment:
I was thinking something similar to this behavior could be tickled if a .lock() request were made while Puppet Server is in the process of flushing an instance that had hit its max-requests counter.  In this case, the old instance would just be unregistered but not put back into the pool.  I was looking to ensure that the .lock() would still succeed after instances had been unregistered.  This test did help me find a reference counting bug in my initial implementation.

I agree with the idea about adding a test for when the last registered instance is unregistered.  I'll do that.